common/src/escrow_accounts.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2023-, GraphOps and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::HashMap, time::Duration};
+
+use alloy_primitives::Address;
+use anyhow::Result;
+use ethers_core::types::U256;
+use eventuals::{timer, Eventual, EventualExt};
+use log::warn;
+use serde::Deserialize;
+use serde_json::json;
+use tokio::time::sleep;
+
+use crate::prelude::SubgraphClient;
+
+pub fn escrow_accounts(
+    escrow_subgraph: &'static SubgraphClient,
+    indexer_address: Address,
+    interval: Duration,
+) -> Eventual<HashMap<Address, U256>> {
+    // Types for deserializing the network subgraph response
+    #[derive(Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    struct EscrowAccountsResponse {
+        escrow_accounts: Vec<EscrowAccount>,
+    }
+    // These 2 structs are used to deserialize the response from the escrow subgraph.
+    // Note that U256's serde implementation is based on serializing the internal bytes, not the string decimal
+    // representation. This is why we deserialize them as strings below.
+    #[derive(Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    struct EscrowAccount {
+        balance: String,
+        total_amount_thawing: String,
+        sender: Sender,
+    }
+    #[derive(Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    struct Sender {
+        id: Address,
+    }
+
+    timer(interval).map_with_retry(
+        move |_| async move {
+            let response = escrow_subgraph
+                .query::<EscrowAccountsResponse>(&json!({
+                    "query": r#"
+                        query ($indexer: ID!) {
+                            escrowAccounts(where: {receiver_: {id: $indexer}}) {
+                                balance
+                                totalAmountThawing
+                                sender {
+                                    id
+                                }
+                            }
+                        }
+                    "#,
+                    "variables": {
+                        "indexer": indexer_address,
+                    }
+                  }
+                ))
+                .await
+                .map_err(|e| e.to_string())?;
+
+            // If there are any GraphQL errors returned, we'll log them for debugging
+            if let Some(errors) = response.errors {
+                warn!(
+                    "Errors encountered fetching escrow accounts for indexer {:?}: {}",
+                    indexer_address,
+                    errors
+                        .into_iter()
+                        .map(|e| e.message)
+                        .collect::<Vec<_>>()
+                        .join(", ")
+                );
```

Comment:
I'd log this as an error at least, since the indexers would refuse service if the query went wrong.

Reply:
The way I think about errors is that you should only log errors if they are severe enough for you to get up at night. Everything else is a warning (or something else that's less severe).

That actually reminds me: whenever we encounter issues like this, we should update a prometheus metric (like indexer-service does with `indexerError`, which both logs the error and updates a metric). Will put that on my todo list.

Reply:
What do you think about my reasoning? I'm ok logging an error if you have a different perspective.

Reply:
Actually, I broke my own rule several times in indexer-service. So let's do an `error!` here like you said.

Reply:
Fixed!

Reply:
> The way I think about errors is that you should only log errors if they are severe enough for you to get up at night. Everything else is a warning (or something else that's less severe).

The indexer-service would immediately start rejecting all queries. IMO that's quite bad.

> That actually reminds me: whenever we encounter issues like this, we should update a prometheus metric (like indexer-service does with indexerError, which both logs the error and updates a metric). Will put that on my todo list.

:100:

Reply:
> > The way I think about errors is that you should only log errors if they are severe enough for you to get up at night. Everything else is a warning (or something else that's less severe).
> 
> The indexer-service would immediately start rejecting all queries. IMO that's quite bad.

Actually, after putting in some more thought, it's slightly more subtle than that, I think. Here it will just log the errors and move on, so at worst it will yield an empty list of accounts:
https://github.com/graphops/indexer-service-rs/blob/b567f9a28d35a674a24d4d0d30cabcfa1c1663ea/common/src/escrow_accounts.rs#L67-L78
That would result in the indexer-service refusing all queries.
Whereas here it will retry a few times, and perhaps just keep the old list of accounts?
https://github.com/graphops/indexer-service-rs/blob/b567f9a28d35a674a24d4d0d30cabcfa1c1663ea/common/src/escrow_accounts.rs#L67-L78
The risk here is that the indexer-service serves queries to escrow accounts that got emptied. But it should get caught by the TAP agent before going over the maximal admissible loss.

Overall, I'm not quite sure what the best solution is here :shrug: